File: windows/ClipSync.Core/Protocol/SyncMessages.cs
using System.Text.Json.Serialization;

namespace ClipSync.Core.Protocol;

public static class ProtocolMessageTypes
{
    public const string Hello = "hello";
    public const string Challenge = "challenge";
    public const string Auth = "auth";
    public const string KnownVector = "known_vector";
    public const string WantRanges = "want_ranges";
    public const string ClipAnnounce = "clip_announce";
    public const string ClipFetch = "clip_fetch";
    public const string ClipPayload = "clip_payload";
    public const string ClipPayloadBegin = "clip_payload_begin";
    public const string ClipPayloadChunk = "clip_payload_chunk";
    public const string ClipPayloadEnd = "clip_payload_end";
    public const string AckRanges = "ack_ranges";
    public const string Error = "error";
    public const string Ping = "ping";
    public const string Pong = "pong";

    public static readonly IReadOnlySet<string> All = new HashSet<string>(StringComparer.Ordinal)
    {
        Hello, Challenge, Auth, KnownVector, WantRanges, ClipAnnounce,
        ClipFetch, ClipPayload, ClipPayloadBegin, ClipPayloadChunk, ClipPayloadEnd,
        AckRanges, Error, Ping, Pong
    };

    public static readonly IReadOnlySet<string> V1 = new HashSet<string>(StringComparer.Ordinal)
    {
        Hello, Challenge, Auth, KnownVector, WantRanges, ClipAnnounce,
        ClipFetch, ClipPayload, AckRanges, Error, Ping, Pong
    };
}

public static class ProtocolLimits
{
    public const int ProtocolVersion = 1;
    public const int ProtocolVersionV2 = 2;
    public const int MaxJsonDepth = 16;
    public const int MaxContentUtf8Bytes = 1_048_576;
    public const int MaxPayloadBatchContentBytes = 1_048_576;
    public const int MaxAnnounceClips = 256;
    public const int MaxPayloadClips = 32;
    public const int MaxFetchEventIds = 128;
    public const int MaxOriginsPerMessage = 128;
    public const int MaxRangesPerOrigin = 256;
    public const int MaxWebSocketTextMessageBytes = 7 * 1_048_576;
    public const int MaxSourceAppLength = 256;
    public const int MaxClientVersionLength = 64;
    public const long MaxRetryAfterMs = 300_000;
    public const int MaxCapabilities = 16;
    public const int MaxEncodedImageBytes = 16 * 1024 * 1024;
    public const int MaxImagePixels = 32 * 1024 * 1024;
    public const int MaxImageSide = 8_192;
    public const int MaxChunkBytes = 256 * 1024;
    public const int MaxChunkCount = 64;
    public const int MaxConcurrentImageDownloads = 2;
    public const string CapabilityImageClipV2 = "image_clip_v2";
}

public static class ProtocolErrorCodes
{
    public const string MalformedJson = "MALFORMED_JSON";
    public const string SchemaViolation = "SCHEMA_VIOLATION";
    public const string UnsupportedVersion = "UNSUPPORTED_VERSION";
    public const string AuthRequired = "AUTH_REQUIRED";
    public const string AuthFailed = "AUTH_FAILED";
    public const string ChallengeExpired = "CHALLENGE_EXPIRED";
    public const string ReplayDetected = "REPLAY_DETECTED";
    public const string DeviceRevoked = "DEVICE_REVOKED";
    public const string TrustEpochMismatch = "TRUST_EPOCH_MISMATCH";
    public const string MessageOutOfOrder = "MESSAGE_OUT_OF_ORDER";
    public const string InvalidRange = "INVALID_RANGE";
    public const string EventConflict = "EVENT_CONFLICT";
    public const string PayloadNotFound = "PAYLOAD_NOT_FOUND";
    public const string HashMismatch = "HASH_MISMATCH";
    public const string PayloadTooLarge = "PAYLOAD_TOO_LARGE";
    public const string RateLimited = "RATE_LIMITED";
    public const string InternalError = "INTERNAL_ERROR";
    public const string UnsupportedMedia = "UNSUPPORTED_MEDIA";
    public const string MediaTooLarge = "MEDIA_TOO_LARGE";
    public const string MediaDecodeFailed = "MEDIA_DECODE_FAILED";
    public const string MediaHashMismatch = "MEDIA_HASH_MISMATCH";
    public const string MediaOutOfOrder = "MEDIA_OUT_OF_ORDER";
    public const string MediaStorageFailed = "MEDIA_STORAGE_FAILED";

    public static readonly IReadOnlySet<string> All = new HashSet<string>(StringComparer.Ordinal)
    {
        MalformedJson, SchemaViolation, UnsupportedVersion, AuthRequired, AuthFailed,
        ChallengeExpired, ReplayDetected, DeviceRevoked, TrustEpochMismatch, MessageOutOfOrder,
        InvalidRange, EventConflict, PayloadNotFound, HashMismatch, PayloadTooLarge,
        RateLimited, InternalError
    };

    public static readonly IReadOnlySet<string> V2 = new HashSet<string>(StringComparer.Ordinal)
    {
        MalformedJson, SchemaViolation, UnsupportedVersion, AuthRequired, AuthFailed,
        ChallengeExpired, ReplayDetected, DeviceRevoked, TrustEpochMismatch, MessageOutOfOrder,
        InvalidRange, EventConflict, PayloadNotFound, HashMismatch, PayloadTooLarge,
        RateLimited, InternalError, UnsupportedMedia, MediaTooLarge, MediaDecodeFailed,
        MediaHashMismatch, MediaOutOfOrder, MediaStorageFailed
    };
}

public sealed record RangeDto
{
    [JsonPropertyName("start_seq")]
    public required long StartSeq { get; init; }

    [JsonPropertyName("end_seq")]
    public required long EndSeq { get; init; }
}

public sealed record OriginStateDto
{
    [JsonPropertyName("origin_device_id")]
    public required string OriginDeviceId { get; init; }

    [JsonPropertyName("contiguous_seq")]
    public required long ContiguousSeq { get; init; }

    [JsonPropertyName("received_ranges")]
    public IReadOnlyList<RangeDto>? ReceivedRanges { get; init; }
}

public sealed record SyncStateDto
{
    [JsonPropertyName("origins")]
    public required IReadOnlyList<OriginStateDto> Origins { get; init; }
}

public sealed record HelloBody
{
    [JsonPropertyName("device_id")]
    public required string DeviceId { get; init; }

    [JsonPropertyName("platform")]
    public required string Platform { get; init; }

    [JsonPropertyName("client_version")]
    public required string ClientVersion { get; init; }

    [JsonPropertyName("trust_epoch")]
    public required long TrustEpoch { get; init; }

    [JsonPropertyName("known_vector")]
    public required SyncStateDto KnownVector { get; init; }

    [JsonPropertyName("capabilities")]
    public IReadOnlyList<string>? Capabilities { get; init; }
}

public sealed record ChallengeBody
{
    [JsonPropertyName("algorithm")]
    public required string Algorithm { get; init; }

    [JsonPropertyName("nonce")]
    public required string Nonce { get; init; }

    [JsonPropertyName("challenger_device_id")]
    public required string ChallengerDeviceId { get; init; }

    [JsonPropertyName("responder_device_id")]
    public required string ResponderDeviceId { get; init; }

    [JsonPropertyName("trust_epoch")]
    public required long TrustEpoch { get; init; }

    [JsonPropertyName("expires_at_ms")]
    public required long ExpiresAtMs { get; init; }
}

public sealed record AuthBody
{
    [JsonPropertyName("algorithm")]
    public required string Algorithm { get; init; }

    [JsonPropertyName("challenge_request_id")]
    public required string ChallengeRequestId { get; init; }

    [JsonPropertyName("responder_device_id")]
    public required string ResponderDeviceId { get; init; }

    [JsonPropertyName("trust_epoch")]
    public required long TrustEpoch { get; init; }

    [JsonPropertyName("proof")]
    public required string Proof { get; init; }
}

public sealed record OriginRangesDto
{
    [JsonPropertyName("origin_device_id")]
    public required string OriginDeviceId { get; init; }

    [JsonPropertyName("ranges")]
    public required IReadOnlyList<RangeDto> Ranges { get; init; }
}

public sealed record WantRangesBody
{
    [JsonPropertyName("requests")]
    public required IReadOnlyList<OriginRangesDto> Requests { get; init; }
}

public static class ClipAvailability
{
    public const string Available = "available";
    public const string Unavailable = "unavailable";
}

public static class ClipUnavailableReasons
{
    public const string LocalOnly = "local_only";
    public const string Deleted = "deleted";
    public const string Expired = "expired";
    public const string PolicyFiltered = "policy_filtered";
    public const string NotFound = "not_found";

    public const string UnsupportedMedia = "unsupported_media";

    public static readonly IReadOnlySet<string> All = new HashSet<string>(StringComparer.Ordinal)
    {
        LocalOnly, Deleted, Expired, PolicyFiltered, NotFound
    };

    public static readonly IReadOnlySet<string> V2 = new HashSet<string>(StringComparer.Ordinal)
    {
        LocalOnly, Deleted, Expired, PolicyFiltered, NotFound, UnsupportedMedia
    };
}

public sealed record ClipHeaderDto
{
    [JsonPropertyName("event_id")]
    public required string EventId { get; init; }

    [JsonPropertyName("origin_device_id")]
    public required string OriginDeviceId { get; init; }

    [JsonPropertyName("origin_seq")]
    public required long OriginSeq { get; init; }

    [JsonPropertyName("availability")]
    public required string Availability { get; init; }

    [JsonPropertyName("kind")]
    public string? Kind { get; init; }

    [JsonPropertyName("content_hash")]
    public string? ContentHash { get; init; }

    [JsonPropertyName("utf8_bytes")]
    public long? Utf8Bytes { get; init; }

    [JsonPropertyName("source_app")]
    public string? SourceApp { get; init; }

    [JsonPropertyName("created_at_ms")]
    public long? CreatedAtMs { get; init; }

    [JsonPropertyName("expires_at_ms")]
    public long? ExpiresAtMs { get; init; }

    [JsonPropertyName("reason")]
    public string? Reason { get; init; }

    [JsonPropertyName("mime_type")]
    public string? MimeType { get; init; }

    [JsonPropertyName("encoded_bytes")]
    public long? EncodedBytes { get; init; }

    [JsonPropertyName("pixel_width")]
    public long? PixelWidth { get; init; }

    [JsonPropertyName("pixel_height")]
    public long? PixelHeight { get; init; }
}

public sealed record ClipAnnounceBody
{
    [JsonPropertyName("clips")]
    public required IReadOnlyList<ClipHeaderDto> Clips { get; init; }
}

public sealed record ClipFetchBody
{
    [JsonPropertyName("event_ids")]
    public required IReadOnlyList<string> EventIds { get; init; }
}

public sealed record ClipPayloadItemDto
{
    [JsonPropertyName("event_id")]
    public required string EventId { get; init; }

    [JsonPropertyName("origin_device_id")]
    public required string OriginDeviceId { get; init; }

    [JsonPropertyName("origin_seq")]
    public required long OriginSeq { get; init; }

    [JsonPropertyName("kind")]
    public required string Kind { get; init; }

    [JsonPropertyName("content")]
    public required string Content { get; init; }

    [JsonPropertyName("content_hash")]
    public required string ContentHash { get; init; }

    [JsonPropertyName("utf8_bytes")]
    public required long Utf8Bytes { get; init; }

    [JsonPropertyName("source_app")]
    public string? SourceApp { get; init; }

    [JsonPropertyName("created_at_ms")]
    public required long CreatedAtMs { get; init; }

    [JsonPropertyName("expires_at_ms")]
    public long? ExpiresAtMs { get; init; }
}

public sealed record ClipPayloadBody
{
    [JsonPropertyName("clips")]
    public required IReadOnlyList<ClipPayloadItemDto> Clips { get; init; }
}

public sealed record ClipPayloadBeginBody
{
    [JsonPropertyName("transfer_id")]
    public required string TransferId { get; init; }

    [JsonPropertyName("event_id")]
    public required string EventId { get; init; }

    [JsonPropertyName("chunk_count")]
    public required long ChunkCount { get; init; }

    [JsonPropertyName("encoded_bytes")]
    public required long EncodedBytes { get; init; }

    [JsonPropertyName("content_hash")]
    public required string ContentHash { get; init; }

    [JsonPropertyName("mime_type")]
    public required string MimeType { get; init; }
}

public sealed record ClipPayloadChunkBody
{
    [JsonPropertyName("transfer_id")]
    public required string TransferId { get; init; }

    [JsonPropertyName("event_id")]
    public required string EventId { get; init; }

    [JsonPropertyName("chunk_index")]
    public required long ChunkIndex { get; init; }

    [JsonPropertyName("chunk_count")]
    public required long ChunkCount { get; init; }

    [JsonPropertyName("chunk_bytes")]
    public required long ChunkBytes { get; init; }

    [JsonPropertyName("data")]
    public required string Data { get; init; }
}

public sealed record ClipPayloadEndBody
{
    [JsonPropertyName("transfer_id")]
    public required string TransferId { get; init; }

    [JsonPropertyName("event_id")]
    public required string EventId { get; init; }

    [JsonPropertyName("content_hash")]
    public required string ContentHash { get; init; }
}

public sealed record AckRangesBody
{
    [JsonPropertyName("acks")]
    public required IReadOnlyList<OriginRangesDto> Acks { get; init; }
}

public sealed record ErrorBody
{
    [JsonPropertyName("code")]
    public required string Code { get; init; }

    [JsonPropertyName("retryable")]
    public required bool Retryable { get; init; }

    [JsonPropertyName("failed_type")]
    public string? FailedType { get; init; }

    [JsonPropertyName("retry_after_ms")]
    public long? RetryAfterMs { get; init; }
}

public sealed record PingBody
{
    [JsonPropertyName("sent_at_ms")]
    public required long SentAtMs { get; init; }
}

public sealed record PongBody
{
    [JsonPropertyName("ping_sent_at_ms")]
    public required long PingSentAtMs { get; init; }

    [JsonPropertyName("sent_at_ms")]
    public required long SentAtMs { get; init; }
}

File: android/app/src/main/java/com/clipsync/android/protocol/SyncMessages.kt
package com.clipsync.android.protocol

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.decodeFromJsonElement

object ProtocolMessageTypes {
    const val HELLO = "hello"
    const val CHALLENGE = "challenge"
    const val AUTH = "auth"
    const val KNOWN_VECTOR = "known_vector"
    const val WANT_RANGES = "want_ranges"
    const val CLIP_ANNOUNCE = "clip_announce"
    const val CLIP_FETCH = "clip_fetch"
    const val CLIP_PAYLOAD = "clip_payload"
    const val CLIP_PAYLOAD_BEGIN = "clip_payload_begin"
    const val CLIP_PAYLOAD_CHUNK = "clip_payload_chunk"
    const val CLIP_PAYLOAD_END = "clip_payload_end"
    const val ACK_RANGES = "ack_ranges"
    const val ERROR = "error"
    const val PING = "ping"
    const val PONG = "pong"

    val ALL = setOf(
        HELLO, CHALLENGE, AUTH, KNOWN_VECTOR, WANT_RANGES, CLIP_ANNOUNCE,
        CLIP_FETCH, CLIP_PAYLOAD, ACK_RANGES, ERROR, PING, PONG,
    )

    val ALL_V2 = ALL + setOf(CLIP_PAYLOAD_BEGIN, CLIP_PAYLOAD_CHUNK, CLIP_PAYLOAD_END)
}

object ProtocolLimits {
    const val PROTOCOL_VERSION = 1
    const val PROTOCOL_VERSION_V2 = 2
    const val MAX_JSON_DEPTH = 16
    const val MAX_CONTENT_UTF8_BYTES = 1_048_576
    const val MAX_PAYLOAD_BATCH_CONTENT_BYTES = 1_048_576
    const val MAX_ANNOUNCE_CLIPS = 256
    const val MAX_PAYLOAD_CLIPS = 32
    const val MAX_FETCH_EVENT_IDS = 128
    const val MAX_ORIGINS_PER_MESSAGE = 128
    const val MAX_RANGES_PER_ORIGIN = 256
    const val MAX_WEBSOCKET_TEXT_MESSAGE_BYTES = 7 * 1_048_576
    const val MAX_SOURCE_APP_LENGTH = 256
    const val MAX_CLIENT_VERSION_LENGTH = 64
    const val MAX_RETRY_AFTER_MS = 300_000L
    const val MAX_CAPABILITIES = 16
    const val MAX_ENCODED_IMAGE_BYTES = 16 * 1024 * 1024
    const val MAX_IMAGE_PIXELS = 32 * 1024 * 1024
    const val MAX_IMAGE_SIDE = 8_192
    const val MAX_CHUNK_BYTES = 256 * 1024
    const val MAX_CHUNK_COUNT = 64
    const val MAX_CONCURRENT_IMAGE_DOWNLOADS = 2
    const val CAPABILITY_IMAGE_CLIP_V2 = "image_clip_v2"
}

object ProtocolErrorCodes {
    const val MALFORMED_JSON = "MALFORMED_JSON"
    const val SCHEMA_VIOLATION = "SCHEMA_VIOLATION"
    const val UNSUPPORTED_VERSION = "UNSUPPORTED_VERSION"
    const val AUTH_REQUIRED = "AUTH_REQUIRED"
    const val AUTH_FAILED = "AUTH_FAILED"
    const val CHALLENGE_EXPIRED = "CHALLENGE_EXPIRED"
    const val REPLAY_DETECTED = "REPLAY_DETECTED"
    const val DEVICE_REVOKED = "DEVICE_REVOKED"
    const val TRUST_EPOCH_MISMATCH = "TRUST_EPOCH_MISMATCH"
    const val MESSAGE_OUT_OF_ORDER = "MESSAGE_OUT_OF_ORDER"
    const val INVALID_RANGE = "INVALID_RANGE"
    const val EVENT_CONFLICT = "EVENT_CONFLICT"
    const val PAYLOAD_NOT_FOUND = "PAYLOAD_NOT_FOUND"
    const val HASH_MISMATCH = "HASH_MISMATCH"
    const val PAYLOAD_TOO_LARGE = "PAYLOAD_TOO_LARGE"
    const val RATE_LIMITED = "RATE_LIMITED"
    const val INTERNAL_ERROR = "INTERNAL_ERROR"
    const val UNSUPPORTED_MEDIA = "UNSUPPORTED_MEDIA"
    const val MEDIA_TOO_LARGE = "MEDIA_TOO_LARGE"
    const val MEDIA_DECODE_FAILED = "MEDIA_DECODE_FAILED"
    const val MEDIA_HASH_MISMATCH = "MEDIA_HASH_MISMATCH"
    const val MEDIA_OUT_OF_ORDER = "MEDIA_OUT_OF_ORDER"
    const val MEDIA_STORAGE_FAILED = "MEDIA_STORAGE_FAILED"

    val ALL = setOf(
        MALFORMED_JSON, SCHEMA_VIOLATION, UNSUPPORTED_VERSION, AUTH_REQUIRED, AUTH_FAILED,
        CHALLENGE_EXPIRED, REPLAY_DETECTED, DEVICE_REVOKED, TRUST_EPOCH_MISMATCH, MESSAGE_OUT_OF_ORDER,
        INVALID_RANGE, EVENT_CONFLICT, PAYLOAD_NOT_FOUND, HASH_MISMATCH, PAYLOAD_TOO_LARGE,
        RATE_LIMITED, INTERNAL_ERROR,
    )

    val ALL_V2 = ALL + setOf(
        UNSUPPORTED_MEDIA, MEDIA_TOO_LARGE, MEDIA_DECODE_FAILED, MEDIA_HASH_MISMATCH,
        MEDIA_OUT_OF_ORDER, MEDIA_STORAGE_FAILED,
    )
}

object ClipAvailability {
    const val AVAILABLE = "available"
    const val UNAVAILABLE = "unavailable"
}

object ClipUnavailableReasons {
    const val LOCAL_ONLY = "local_only"
    const val DELETED = "deleted"
    const val EXPIRED = "expired"
    const val POLICY_FILTERED = "policy_filtered"
    const val NOT_FOUND = "not_found"
    const val UNSUPPORTED_MEDIA = "unsupported_media"

    val ALL = setOf(LOCAL_ONLY, DELETED, EXPIRED, POLICY_FILTERED, NOT_FOUND)
    val ALL_V2 = ALL + setOf(UNSUPPORTED_MEDIA)
}

sealed interface SyncMessageBody

@Serializable
data class RangeDto(
    @SerialName("start_seq") val startSeq: Long,
    @SerialName("end_seq") val endSeq: Long,
)

@Serializable
data class OriginStateDto(
    @SerialName("origin_device_id") val originDeviceId: String,
    @SerialName("contiguous_seq") val contiguousSeq: Long,
    @SerialName("received_ranges") val receivedRanges: List<RangeDto>? = null,
)

@Serializable
data class SyncStateDto(
    val origins: List<OriginStateDto>,
) : SyncMessageBody

@Serializable
data class HelloBody(
    @SerialName("device_id") val deviceId: String,
    val platform: String,
    @SerialName("client_version") val clientVersion: String,
    @SerialName("trust_epoch") val trustEpoch: Long,
    @SerialName("known_vector") val knownVector: SyncStateDto,
    val capabilities: List<String>? = null,
) : SyncMessageBody

@Serializable
data class ChallengeBody(
    val algorithm: String,
    val nonce: String,
    @SerialName("challenger_device_id") val challengerDeviceId: String,
    @SerialName("responder_device_id") val responderDeviceId: String,
    @SerialName("trust_epoch") val trustEpoch: Long,
    @SerialName("expires_at_ms") val expiresAtMs: Long,
) : SyncMessageBody

@Serializable
data class AuthBody(
    val algorithm: String,
    @SerialName("challenge_request_id") val challengeRequestId: String,
    @SerialName("responder_device_id") val responderDeviceId: String,
    @SerialName("trust_epoch") val trustEpoch: Long,
    val proof: String,
) : SyncMessageBody

@Serializable
data class OriginRangesDto(
    @SerialName("origin_device_id") val originDeviceId: String,
    val ranges: List<RangeDto>,
)

@Serializable
data class WantRangesBody(
    val requests: List<OriginRangesDto>,
) : SyncMessageBody

@Serializable
data class ClipHeaderDto(
    @SerialName("event_id") val eventId: String,
    @SerialName("origin_device_id") val originDeviceId: String,
    @SerialName("origin_seq") val originSeq: Long,
    val availability: String,
    val kind: String? = null,
    @SerialName("content_hash") val contentHash: String? = null,
    @SerialName("utf8_bytes") val utf8Bytes: Long? = null,
    @SerialName("source_app") val sourceApp: String? = null,
    @SerialName("created_at_ms") val createdAtMs: Long? = null,
    @SerialName("expires_at_ms") val expiresAtMs: Long? = null,
    val reason: String? = null,
    @SerialName("mime_type") val mimeType: String? = null,
    @SerialName("encoded_bytes") val encodedBytes: Long? = null,
    @SerialName("pixel_width") val pixelWidth: Long? = null,
    @SerialName("pixel_height") val pixelHeight: Long? = null,
)

@Serializable
data class ClipAnnounceBody(
    val clips: List<ClipHeaderDto>,
) : SyncMessageBody

@Serializable
data class ClipFetchBody(
    @SerialName("event_ids") val eventIds: List<String>,
) : SyncMessageBody

@Serializable
data class ClipPayloadItemDto(
    @SerialName("event_id") val eventId: String,
    @SerialName("origin_device_id") val originDeviceId: String,
    @SerialName("origin_seq") val originSeq: Long,
    val kind: String,
    val content: String,
    @SerialName("content_hash") val contentHash: String,
    @SerialName("utf8_bytes") val utf8Bytes: Long,
    @SerialName("source_app") val sourceApp: String? = null,
    @SerialName("created_at_ms") val createdAtMs: Long,
    @SerialName("expires_at_ms") val expiresAtMs: Long? = null,
)

@Serializable
data class ClipPayloadBody(
    val clips: List<ClipPayloadItemDto>,
) : SyncMessageBody

@Serializable
data class ClipPayloadBeginBody(
    @SerialName("transfer_id") val transferId: String,
    @SerialName("event_id") val eventId: String,
    @SerialName("chunk_count") val chunkCount: Long,
    @SerialName("encoded_bytes") val encodedBytes: Long,
    @SerialName("content_hash") val contentHash: String,
    @SerialName("mime_type") val mimeType: String,
) : SyncMessageBody

@Serializable
data class ClipPayloadChunkBody(
    @SerialName("transfer_id") val transferId: String,
    @SerialName("event_id") val eventId: String,
    @SerialName("chunk_index") val chunkIndex: Long,
    @SerialName("chunk_count") val chunkCount: Long,
    @SerialName("chunk_bytes") val chunkBytes: Long,
    val data: String,
) : SyncMessageBody

@Serializable
data class ClipPayloadEndBody(
    @SerialName("transfer_id") val transferId: String,
    @SerialName("event_id") val eventId: String,
    @SerialName("content_hash") val contentHash: String,
) : SyncMessageBody

@Serializable
data class AckRangesBody(
    val acks: List<OriginRangesDto>,
) : SyncMessageBody

@Serializable
data class ErrorBody(
    val code: String,
    val retryable: Boolean,
    @SerialName("failed_type") val failedType: String? = null,
    @SerialName("retry_after_ms") val retryAfterMs: Long? = null,
) : SyncMessageBody

@Serializable
data class PingBody(
    @SerialName("sent_at_ms") val sentAtMs: Long,
) : SyncMessageBody

@Serializable
data class PongBody(
    @SerialName("ping_sent_at_ms") val pingSentAtMs: Long,
    @SerialName("sent_at_ms") val sentAtMs: Long,
) : SyncMessageBody

data class ParsedSyncMessage(
    val version: Int,
    val type: String,
    val requestId: String,
    val body: SyncMessageBody,
)

/**
 * Typed v1 bodies. Envelope JSON is validated by [ProtocolJson.parseEnvelope]; this
 * object only maps the already-accepted body object onto Kotlin types.
 */
object SyncMessages {
    private val json = Json {
        ignoreUnknownKeys = false
        isLenient = false
        coerceInputValues = false
        explicitNulls = false
    }

    fun parse(source: String): ParsedSyncMessage =
        parse(source, ProtocolLimits.PROTOCOL_VERSION)

    fun parse(source: String, version: Int): ParsedSyncMessage {
        val envelope = if (version == ProtocolLimits.PROTOCOL_VERSION_V2) {
            ProtocolJson.parseEnvelopeV2(source)
        } else {
            ProtocolJson.parseEnvelope(source)
        }
        return ParsedSyncMessage(
            version = envelope.version,
            type = envelope.type,
            requestId = envelope.requestId,
            body = decodeBody(envelope.type, envelope.body),
        )
    }

    fun decodeBody(type: String, body: JsonObject): SyncMessageBody = when (type) {
        ProtocolMessageTypes.HELLO -> json.decodeFromJsonElement<HelloBody>(body)
        ProtocolMessageTypes.CHALLENGE -> json.decodeFromJsonElement<ChallengeBody>(body)
        ProtocolMessageTypes.AUTH -> json.decodeFromJsonElement<AuthBody>(body)
        ProtocolMessageTypes.KNOWN_VECTOR -> json.decodeFromJsonElement<SyncStateDto>(body)
        ProtocolMessageTypes.WANT_RANGES -> json.decodeFromJsonElement<WantRangesBody>(body)
        ProtocolMessageTypes.CLIP_ANNOUNCE -> json.decodeFromJsonElement<ClipAnnounceBody>(body)
        ProtocolMessageTypes.CLIP_FETCH -> json.decodeFromJsonElement<ClipFetchBody>(body)
        ProtocolMessageTypes.CLIP_PAYLOAD -> json.decodeFromJsonElement<ClipPayloadBody>(body)
        ProtocolMessageTypes.CLIP_PAYLOAD_BEGIN -> json.decodeFromJsonElement<ClipPayloadBeginBody>(body)
        ProtocolMessageTypes.CLIP_PAYLOAD_CHUNK -> json.decodeFromJsonElement<ClipPayloadChunkBody>(body)
        ProtocolMessageTypes.CLIP_PAYLOAD_END -> json.decodeFromJsonElement<ClipPayloadEndBody>(body)
        ProtocolMessageTypes.ACK_RANGES -> json.decodeFromJsonElement<AckRangesBody>(body)
        ProtocolMessageTypes.ERROR -> json.decodeFromJsonElement<ErrorBody>(body)
        ProtocolMessageTypes.PING -> json.decodeFromJsonElement<PingBody>(body)
        ProtocolMessageTypes.PONG -> json.decodeFromJsonElement<PongBody>(body)
        else -> error("Unsupported message type: $type")
    }
}
